### DISCUSSION TITLE
xDrip+ goes into a crash/reboot loop if Google Play Services is disabled with Google Health Connect enabled

### DISCUSSION THREAD

**Discussion by jbowler** [NightscoutFoundation/xDrip#2808] [General]
The precise problem seems to be this:

```
HealthGamut
Companion app not available - asking for installation
```

HealthGamut seems to be associated with Google Health Connect; if I turn it off I instead get a warning about "Google Play Services" being disabled.

**Comment by Navid200** [comment]
According to this, you cannot use Google Health connect without Google Play service:
https://developer.android.com/guide/health-and-fitness/health-connect/frequently-asked-questions

![Screenshot 2023-05-02 112421](https://user-images.githubusercontent.com/51497406/235711881-ecdf993f-af29-485a-a8fd-e9682d6855e3.png)

If that is the case, there would be no working scenario under those conditions.  So, I'm not sure we can even consider changing xDrip to address a scenario that we know is going to fail by design.

Please let me know if I have misunderstood you.

**Reply by jbowler** [reply]
No, that's not what I meant.  The problem is that xDrip+ goes into an infinite crash/reboot loop.  I don't know if that is because of HealthConnect or, for that matter, HealthGamut, I just know that the bug "disappeared" if I disabled HealthConnect.  I.e. to be clear there is a bug in xDrip+ which causes an infinite crash/reboot loop and the rest of what I said is the repro I have and all the information that I thought _might_ be relevant.

**Reply by Navid200** [reply]
Can you tell me how I can recreate this reboot loop on my device?

**Reply by jbowler** [reply]
My repro is on an Android 13 device.  When I found the bug I had uninstalled everything I could except xDrip+ and AndroidAPS and I had also disabled everything I could; not services, but all the apps, including the hidden ones, that I could find.  It isn't necessary to do that.  Basic setup:

Android 13, March 5, 2023 security update
xDrip+ 2123042419 [original repro was an earlier nightly build]
Google Health stuff **NOT** installed
Running, configured, xDrip+

Steps.  **NOTE**: I'm giving steps for Android 13, the Android GUI changes with every new day, but I suspect the repro will work with older versions too so long as the Android bits are adjusted.

1. xDrip+/hamburger/Settings/Inter-app settings
2. Click "Google Health Connect" (at the bottom)
3. Click "Use Health Connect" (turns the check mark from "off" to "on").
4. _Observe_: Google Play Store appears with a window to "Install" Health Connect (Beta).  _Do not do this -)_
5. Return to the 'phone home screen.
6. Settings/Apps/App management
7. Scroll down to "Google Play services", click on it.
8. Click "Disable", click "Disable app" and disable the app (password required)
9. Go back, "App info", scroll down some more to "Google Play Store" and click on it.
10. Click "Disable", "Disable app", to disable that too.
11. **Reboot the 'phone**; typically "Power button"/restart
12. Click on xDrip+ restart it
13. _Observe_: the 'phone pops up a message saying "xDrip+ keeps stopping"

Since anyone trying this is going to want to undo it the "undo" is to go to "settings" and reenable "Google Play Services" and "Google Play Store" then reboot the 'phone, launch xDrip and _as fast as you_ can and disable "Google Health Connect" (as in stesp (2) and (3) above); it can be challenging but it does work.

To people who are not accustomed to software development; **do not do this on a live system**.

**Reply by Navid200** [reply]
I don't understand.
You are telling me to enable Google Health connect, and disable Google Play service.
But, that combination is not going to work as Google explains in the link I provided above.


If anyone does it and xDrip goes into a reboot loop, just enable Google Play Service to stop the crash loop.
I can add this to the FAQ in documentation.  Would that be good?

Please help me understand.  
I cannot ask developers to change xDrip if I don't understand what the problem is.  I don't see how we can change xDrip to allow Health connect to be active and not constantly looking for Google Play service.

**Reply by jbowler** [reply]
> I don't understand.

I'm sorry; this is just something we do.  It is a "repro" of a "crash".  In software development the latter is a major issue and the former is the holy grail.  You should upstream this; this is basic software test because crashes are unacceptable in apps.

**Reply by Navid200** [reply]
> this is just something we do

I understand that part.
Can you tell me something?
In the scenario you are describing, after the phone reboots, if you don't start xDrip and instead enable Google Play service, will xDrip then not open without a crash?

**Reply by jbowler** [reply]
> if you don't start xDrip and instead enable Google Play service, will xDrip then not open without a crash?

I'm not volunteering to be your STE.  You are asking about a completely different scenario; if you want to know try it!

**Reply by Navid200** [reply]
No,
I was talking about the scenario you have described.

But, that's OK.
I did my best to understand this and channel it to the right developer.

We can leave it and developers may look at it.

**Comment by digiexchris** [comment]
I think I can clarify this. The issue is not that the user is trying to use Health Connect without Play Services. 

The issue is that xdrip crashes if they select Health Connect in the xdrip menu if health connect is not installed and play services is disabled. 

The expected behaviour is for xdrip to either show a warning that health connect is unusable if selected, or block enabling of that feature, instead of crashing.

Also @jbowler  easy there, don't be too hard on them. Everyone working on this is volunteers and no one is getting paid. They depend on people like you and me to do testing and discover more information that helps them solve the problem. Your attitude would not be acceptable in the software company I work for. QC and I have discussions and I try things that they ask about, not just firing a report off into the void and saying it's a them problem.

**Reply by jbowler** [reply]
> The issue is that xdrip crashes if they select Health Connect in the xdrip menu

Not quite; that does work in that if Health Connect is not enabled in xDrip+ I can boot xDrip+ with the warning and I think it works ok (without health connect) but in any case it gives me the warning.

The problem arose because I had previously enabled "Google Health Connect" in "Inter-app settings" while Google Play Services was enabled then I disabled Google Play Services and experienced the problem; xDrip+ won't even boot and it goes into an infinite-crash loop (until the OS gets fed up and disables the auto-restart).

**Reply by digiexchris** [reply]
Ah yes, that does help clarify. This may be a crash with xdrip asking google play services if Health Connect is installed starting up the health connect features.

May I ask why you turn off google play services?

**Reply by jbowler** [reply]
> May I ask why you turn off google play services?

You answered your own question, sort of, :-):

https://github.com/NightscoutFoundation/xDrip/discussions/2809

The only way that I could find in Android 13 to work out which app was making unexpected sounds in the media channel was to uninstall every app until the sound stopped.  Since my sole purpose of having a relatively expensive cell phone is to run AndroidAPS (as a closed loop system) and xDrip+ (as a CGM data collector for AAPS) I uninstalled everything except those two apps.  I seriously don't care if I have to uninstall Google stuff to have a working pancreas :-)

So I got down to the point where I knew one of those two apps was causing the problem (a show stopper for me) then ends up getting the answer as to how to fix the problem on FB (AndroidAPS).  Note that until I had determined that the problem was in one of the two apps it would not have been reasonable to ask what causes my phone to make a loud trumpet noise every few minutes, at least not on FB/AndroidAPS and elsewhere I would not have got the right answer.
